### PR TITLE
Consume the newly added analyzer APIs in all our analyzers

### PR DESCRIPTION
**Pull request by mavasani**
1. EnableConcurrentExecution - few analyzers which are not thread-safe have been marked as consider for making thread-safe.
2. ConfigureGeneratedCodeAnalysis - most analyzers have been configured to not run or report diagnostics on generated code - except for security analyzers and DeclarePublicAPI analyzer (Analyze and ReportDiagnostics in generated code) and compilation end analyzers (Analyze but do not report diagnostics in generated code).

I plan to build NuGet packages with this change, and run it on some RWC code (maybe just move Analyzers.sln to newer analyzers) to confirm no regressions.

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet/roslyn-analysis @srivatsn @basoundr can you please review?

File: src/Desktop.Analyzers/Core/DoNotUseInsecureCryptographicAlgorithms.cs [hunk 1]
```diff
@@ -55,6 +55,11 @@ private static DiagnosticDescriptor CreateDiagnosticDescriptor(string ruleId, Lo
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics in generated code.
```
Comment (heejaechang) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureCryptographicAlgorithms.cs]: for generated code, once user has diagnostic reported for the file, except suppressing it, what is users options to fix the diagnostic?

Comment (heejaechang) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureCryptographicAlgorithms.cs]: is idea that user stop using tool that generate such code?

File: src/Desktop.Analyzers/Core/TypesShouldNotExtendCertainBaseTypes.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ public sealed class TypesShouldNotExtendCertainBaseTypesAnalyzer : DiagnosticAna
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
-        private static readonly Dictionary<string, string> s_badBaseTypesToMessage = new Dictionary<string, string>
+        private static readonly ImmutableDictionary<string, string> s_badBaseTypesToMessage = new Dictionary<string, string>
```
Comment (heejaechang) [inline src/Desktop.Analyzers/Core/TypesShouldNotExtendCertainBaseTypes.cs]: this seems compilation error

File: src/System.Runtime.Analyzers/Core/DoNotRaiseReservedExceptionTypes.cs [hunk 3]
```diff
@@ -82,27 +82,30 @@ public abstract class DoNotRaiseReservedExceptionTypesAnalyzer<TLanguageKindEnum
...
-                compilationStartContext.RegisterSyntaxNodeAction(
-                syntaxNodeContext =>
+            analysisContext.RegisterCompilationStartAction(
+                compilationStartContext =>
                 {
-                    Analyze(syntaxNodeContext, tooGenericExceptionSymbols, reservedExceptionSymbols);
-                },
-                ObjectCreationExpressionKind);
-            });
+                    ImmutableHashSet<INamedTypeSymbol> tooGenericExceptionSymbols = CreateSymbolSet(compilationStartContext.Compilation, s_tooGenericExceptions);
```
Comment (heejaechang) [inline src/System.Runtime.Analyzers/Core/DoNotRaiseReservedExceptionTypes.cs]: var?

**Comment by heejaechang** [comment]
👍

File: src/Desktop.Analyzers/Core/DoNotUseInsecureCryptographicAlgorithms.cs (hunk 1 above)
Comment (mavasani) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureCryptographicAlgorithms.cs]: End user will probably suppress the diagnostic as it is in generated code, but will most likely report it to the tool author to fix their code generator.

**Comment by mavasani** [comment]
FYI: I have commented out these invocations in the stub analyzers.

**Comment by srivatsn** [comment]
:+1:

**Comment by mavasani** [comment]
Jenkins seem to be down - tests pass locally and even passed on my previous commit (latest commit just comments out invocations in stub analyzers). I am merging this in.

**Pull request merged**